### DISCUSSION TITLE
Token Revocation

### DISCUSSION THREAD

**Discussion by girishk-btc** [keycloak/keycloak#22776] [Q&A]
Hi,
I need to achieve restrict multi-device login. Is there any way to invalidate all the tokens issued to particular user through rest api in keycloak 21.1.1

**Comment by justin-tay** [comment]
It might be easier to use [User session limits](https://www.keycloak.org/docs/latest/server_admin/index.html#_user_session_limits) with `Deny new session` with a concurrent limit of 1 to enforce this. You will need to add it to all the flows you support.

Unless you are using those tokens to access Keycloak services itself like the userinfo endpoint, when it comes to revoking tokens, in the first place do your services call Keycloak's introspection endpoint to check if tokens are revoked? If not then those access tokens will still be valid at your services until they expire as they wouldn't know it has been revoked.

If you really need to revoke tokens, you can use `Terminate oldest session` with the User session limits, which would lead to those tokens issued for that session becoming invalid, but again you would need to introspect those tokens to find out they are not active.

**Comment by girishk-btc** [comment]
Can you please suggest how to Terminate oldest session for a particular user

**Reply by justin-tay** [reply]
You will need to add the `User session count limiter` step to your authentication flows as documented at https://www.keycloak.org/docs/latest/server_admin/index.html#_user_session_limits.

![sessionlimit](https://github.com/keycloak/keycloak/assets/49700559/6055971b-ccf9-44a5-a791-a44c25b56da4)

**Comment by girishk-btc** [comment]
It works but only thing is it is invalidating refresh token not the access token.
Any Idea how to achieve this

**Reply by justin-tay** [reply]
How are you validating the access token? Are you calling the introspection endpoint?

**Comment by girishk-btc** [comment]
No i am validating through the jwt

**Reply by justin-tay** [reply]
As stated in my previous replies, that wouldn't work. The JWT token is self contained. If you are just validating the JWT token, the only thing that is relevant is the expiration time claim. Keycloak can't update a JWT token that you are already passing around. Only Keycloak knows that the access token has been revoked so you would need to call the introspection endpoint to check.

**Comment by girishk-btc** [comment]
I need to restrict multiple device login. In order to achieve this i need to invalidate both access token and refresh token
With the approach you mentioned above, i am only able to invalidate refresh token

**Reply by justin-tay** [reply]
The approach that I recommended was to use User session limits with `Deny new session` with a concurrent limit of 1.

It isn't that the access token isn't revoked, you need to check with Keycloak at the introspection endpoint to see if it is revoked. In the refresh token case it is because you need to present the refresh token to the Keycloak token endpoint to refresh so you think that only the refresh token was revoked.

**Reply by girishk-btc** [reply]
I have list of api's in my application where in i need to pass authorization as header (i.e access token). Please suggest me how can i restrict the user to access those api's when he is logged in with new device

**Reply by justin-tay** [reply]
If you `Deny new session` with a limit of 1 he would not be able to login with a new device unless he logs out of the old one.

If you want `Terminate oldest session`, the all your services must call Keycloak's introspection endpoint with the access token for all requests to validate the access token. You cannot just validate the JWT token.

**Reply by cgeorgilakis** [reply]
POST /realms/{realm-name}/protocol/openid-connect/revoke
Used for OAuth 2.0 Token Revocation described in [RFC7009](https://datatracker.ietf.org/doc/html/rfc7009).

If you give access token, access token will be revoked.

**Reply by girishk-btc** [reply]
I don't want to specifically revoke by passing particular token, as per my architecture level im not storing any of the tokens.
Is there any way that I can revoke all the tokens issued to a particular user by passing user id
